pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -0,0 +1,447 @@
+		delete(existingDriverMap, csiDriverName)
+		if len(existingDriverMap) == 0 {
+			delete(node.ObjectMeta.Annotations, annotationKeyNodeID)
+		} else {
+			jsonObj, err := json.Marshal(existingDriverMap)
+			if err != nil {
+				return nil, false, fmt.Errorf(
+					"failed while trying to remove key %q from node %q annotation. Existing data: %v",
+					csiDriverName,
+					annotationKeyNodeID,
+					previousAnnotationValue)
+			}
+
+			node.ObjectMeta.Annotations[annotationKeyNodeID] = string(jsonObj)
+		}
+
+		return node, true, nil
+	}
+}
+
+// updateTopologyLabels returns a function that updates labels of a Node object with the given
+// topology information.
+func updateTopologyLabels(topology *csipb.Topology) nodeUpdateFunc {
+	// TODO (verult) do we want to reverse the prefix?
+	return func(node *v1.Node) (*v1.Node, bool, error) {
+		if len(topology.Segments) == 0 {
+			return node, false, nil
+		}
+
+		for k, v := range topology.Segments {
+			if curVal, exists := node.Labels[k]; exists && curVal != v {
+				return nil, false, fmt.Errorf("detected topology value collision: driver reported %q:%q but existing label is %q:%q", k, v, k, curVal)
+			}
+		}
+
+		if node.Labels == nil {
+			node.Labels = make(map[string]string)
+		}
+		for k, v := range topology.Segments {
+			node.Labels[k] = v
+		}
+		return node, true, nil
+	}
+}
+
+func (nim *nodeInfoManager) updateCSINodeInfo(
+	driverName string,
+	driverNodeID string,
+	topology *csipb.Topology) error {
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		nodeInfo, err := nim.csiKubeClient.CsiV1alpha1().CSINodeInfos().Get(string(nim.nodeName), metav1.GetOptions{})
+		if nodeInfo == nil || errors.IsNotFound(err) {
+			return nim.createNodeInfoObject(driverName, driverNodeID, topology)
+		}
+		if err != nil {
+			return err // do not wrap error
+		}
+
+		return nim.updateNodeInfoObject(nodeInfo, driverName, driverNodeID, topology)
+	})
+	if retryErr != nil {
+		return fmt.Errorf("CSINodeInfo update failed: %v", retryErr)
+	}
+	return nil
+}
+
+func (nim *nodeInfoManager) createNodeInfoObject(
+	driverName string,
+	driverNodeID string,
+	topology *csipb.Topology) error {
+
+	var topologyKeys []string
+	for k := range topology.Segments {
+		topologyKeys = append(topologyKeys, k)
+	}
+
+	node, err := nim.k8s.CoreV1().Nodes().Get(string(nim.nodeName), metav1.GetOptions{})
+	if err != nil {
+		return err // do not wrap error
+	}
+
+	nodeIDsFromAnnotation, err := buildNodeIDMapFromAnnotation(node)
+	if err != nil {
+		return err
+	}
+
+	// Construct driver info list from both existing node annotation and input
+	var driverInfos []csiv1alpha1.CSIDriverInfo
+	for n, id := range nodeIDsFromAnnotation {
+		driverInfo := csiv1alpha1.CSIDriverInfo{
+			Driver: n,
+			NodeID: id,
+		}
+		if n == driverName {
+			// Use given information
+			driverInfo.NodeID = driverNodeID
+			driverInfo.TopologyKeys = topologyKeys
+		}
+		driverInfos = append(driverInfos, driverInfo)
```

Comment:
Should we be adding all drivers in this call to just update a single driver?  Do all the CSI drivers register again when we upgrade the node?

Should we be looking at the annotations at all?  IIUC, this is supposed to be the replacement for the annotations

Reply:
Yeah CSI drivers should register again upon node upgrade, and this is here as a safeguard to ensure annotations and CSINodeInfo are consistent while we support both. There is always a chance that kubelet is upgraded while the previous Node object stays, and annotations and CSINodeInfo could be momentarily out of sync.

Reply:
Is it necessary to keep the annotations and CSIDriverInfo in sync on the first driver's registration, if the other driver will register again later?  In an upgrade scenario, then this means we're on the first boot after upgrade and we're processing the first driver registration.  We're adding an entry for all the other drivers too even though they have not gone through registration yet.  Could that cause other issues?

Reply:
I think this method should get whole list of registered CSI drivers so it creates/updates complete CSINodeInfo with all drivers. The result should be the same in the end, however it feels more robust. I don't line CSINodeInfo with empty topology for some drivers just because we translated it from an annotation.

Reply:
This method gets called for a single driver registration though, and only on the first one since it's creating the object for the first time.

Reply:
Discussed with @msau42 offline. The current implementation doesn't guarantee the node info is always up-to-date. In particular, with the current kubelet plugin registration design, there is no way to remove a non-existent driver unless it detects that a driver socket is removed (after unregistration is implemented). If a driver is removed while kubelet is down, the node information will remain. So whatever issue comes up when CSINodeInfo is kept in sync with annotation is unavoidable today.

But I'll remove the logic to keep the two sources in sync, because it complicates the implementation without adding too much benefit. One minor advantage of keeping them in sync is to avoid debugging confusion, but I think it'll be OK.